### ISSUE TITLE
CRF predicts only majority class

### ISSUE THREAD

**Issue by Horsmann** [haifengl/smile#292]
Hi,

I am experimenting with an own prototype part-of-speech tagger that is implemented using  the `CRF` implementation provided in `smile-core`.

I noticed that once I add word-context as feature values the CRF always outputs the majority class of the labels. Without word context (for the dummy data set and the poor features that I added) I reach a substantial higher result. This is not a problem of the data amount. I tried providing a lot more data but with the current word as feature, performance sticks to majority class prediction.

I have a minimal working example below. Setting the variable `useCurrentWord` to `true` adds a feature that provides the current word-form as feature value. This will lead to an accuracy of 8% i.e. majority class assigned (noun). Setting this value to `false` reaches 64%. (this is not a matter of too few training data).

I am new to `smile`, so maybe I am missing something very obvious. I am grateful for pointers.

```
package crfProto;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import smile.data.Attribute;
import smile.data.NominalAttribute;
import smile.data.StringAttribute;
import smile.sequence.CRF;

public class MinimalWorkingExample {

	static List<List<String>> rawData = new ArrayList<>();

	static void loadData () {
		List<String> seq1 = new ArrayList<>();
		seq1.add("Biologist NN");
		seq1.add("Nick NP");
		seq1.add("Bos NP");
		seq1.add("tells VVZ");
		seq1.add("Wikinews NP");
		seq1.add("about IN");
		seq1.add("' ``");
		seq1.add("self-medicating VVG");
		seq1.add("' \"");
		seq1.add("ants NNS");
		seq1.add(". SENT");

		List<String> seq2 = new ArrayList<>();
		seq2.add("In IN");
		seq2.add("a DT");
		seq2.add("recently-accepted JJ");
		seq2.add("Evolution NP");
		seq2.add("paper NN");
		seq2.add("Bos NP");
		seq2.add("and CC");
		seq2.add("colleagues NNS");
		seq2.add("describe VVP");
		seq2.add("ants NNS");
		seq2.add("appearing VVG");
		seq2.add("to TO");
		seq2.add("self-medicate VV");
		seq2.add(". SENT");

		rawData.add(seq1);
		rawData.add(seq2);
	}

	static Map<String, Integer> labelMap = new HashMap<>();
	static boolean useCurrentWord = false; //Set to "true" - prediction collapses to majority class
	public static void main(String[] args) throws ParseException {
		
		loadData();

		List<List<List<String>>> allSeqFeatures = new ArrayList<>();
		List<List<Integer>> allSeqLabels = new ArrayList<>();
		
		//iterate data
		for (List<String> seq : rawData) {
			List<List<String>> seqFeatures = new ArrayList<>();
			List<Integer> labels = new ArrayList<>();

			// for each word/Pos entry
			for (String l : seq) {
				String[] split = l.split(" ");
				String sWord = split[0];
				String sPos = split[1];

				int iPos = getInt(labelMap, sPos);
				labels.add(iPos);
				List<String> features = getFeaturesValues(sWord);
				seqFeatures.add(features);
			}
			allSeqFeatures.add(seqFeatures);
			allSeqLabels.add(labels);
		}
		
		//Features as Attribute type
		List<Attribute> attributes = new ArrayList<>();
		if(useCurrentWord){
			attributes.add(new StringAttribute("w0"));
		}
		attributes.add(new NominalAttribute("ed", new String[] { "1.0", "0.0" }));
		attributes.add(new NominalAttribute("ing", new String[] { "1.0", "0.0" }));
		attributes.add(new NominalAttribute("ion", new String[] { "1.0", "0.0" }));
		attributes.add(new NominalAttribute("ly", new String[] { "1.0", "0.0" }));

		CRF.Trainer t = new CRF.Trainer(attributes.toArray(new Attribute[0]), labelMap.keySet().size());

		double[][][] s = toDoublePrimitive(allSeqFeatures, attributes.toArray(new Attribute[0]));
		int[][] l = toIntegerPrimitive(allSeqLabels);
		CRF crf = t.train(s, l);

		Map<Integer, String> inverseLabel = new HashMap<>();
		for (Map.Entry<String, Integer> entry : labelMap.entrySet()) {
			inverseLabel.put(entry.getValue(), entry.getKey());
		}

		double c = 0;
		double ic = 0;
		for (int i = 0; i < s.length; i++) {
			int[] gl = l[i];
			double[][] d = s[i];
			int[] predict = crf.predict(d);

			for (int j = 0; j < predict.length; j++) {
				if (gl[j] == predict[j]) {
					c++;
				} else {
					ic++;
				}
				System.out.println(inverseLabel.get(gl[j]) + " " + inverseLabel.get(predict[j]));
			}
			
		}

		System.out.println("Acc: " + (c / (c + ic)) * 100);

	}

	private static List<String> getFeaturesValues(String sWord) {

		List<String> features = new ArrayList<>();
		
		if(useCurrentWord){
			features.add(sWord);
		}

		/* primitive word-suffix features */

		if (sWord.endsWith("ed")) {
			features.add("1.0");
		} else {
			features.add("0.0");
		}

		if (sWord.endsWith("ing")) {
			features.add("1.0");
		} else {
			features.add("0.0");
		}

		if (sWord.endsWith("ion")) {
			features.add("1.0");
		} else {
			features.add("0.0");
		}

		if (sWord.endsWith("ly")) {
			features.add("1.0");
		} else {
			features.add("0.0");
		}

		return features;
	}

	private static int getInt(Map<String, Integer> m, String k) {
		if (m.containsKey(k)) {
			return m.get(k);
		}

		int i = m.values().size();
		m.put(k, i);

		return i;
	}

	private static int[][] toIntegerPrimitive(List<List<Integer>> input) {

		int[][] out = new int[input.size()][];
		for (int i = 0; i < input.size(); i++) {
			out[i] = new int[input.get(i).size()];
			for (int j = 0; j < input.get(i).size(); j++) {
				out[i][j] = input.get(i).get(j);
			}
		}

		return out;
	}

	private static double[][][] toDoublePrimitive(List<List<List<String>>> input, Attribute[] att)
			throws ParseException {

		double[][][] out = new double[input.size()][][];
		for (int i = 0; i < input.size(); i++) {
			out[i] = new double[input.get(i).size()][];
			for (int j = 0; j < input.get(i).size(); j++) {
				out[i][j] = new double[input.get(i).get(j).size()];
				for (int k = 0; k < input.get(i).get(j).size(); k++) {
					out[i][j][k] = att[k].valueOf(input.get(i).get(j).get(k));
				}
			}
		}

		return out;
	}
}
```

**Comment by ghostdogpr** [comment]
Hi,

I ran into the same issue.
From what I observed, the problem comes the StringAttribute. As soon as you include one in the list of Attributes, the predictions make no sense (majority rule as you observed).

As a workaround, I used a NominalAttribute and gave the whole list of strings in my training data as values. It works well, but I'm not sure this is intended (if it is, it would be nice that the documentation mention it).
